src/projection.jl

```diff
@@ -135,13 +135,41 @@ ProjectTo(::Real) = ProjectTo{Real}()
 ProjectTo(::Complex) = ProjectTo{Complex}()
 ProjectTo(::Number) = ProjectTo{Number}()
 for T in (Float16, Float32, Float64, ComplexF16, ComplexF32, ComplexF64)
-    # Preserve low-precision floats as accidental promotion is a common perforance bug
+    # Preserve low-precision floats as accidental promotion is a common performance bug
     @eval ProjectTo(::$T) = ProjectTo{$T}()
 end
 ProjectTo(x::Integer) = ProjectTo(float(x))
 ProjectTo(x::Complex{<:Integer}) = ProjectTo(float(x))
-(::ProjectTo{T})(dx::Number) where {T<:Number} = convert(T, dx)
-(::ProjectTo{T})(dx::Number) where {T<:Real} = convert(T, real(dx))
+
+# Preserve low-precision floats as accidental promotion is a common performance bug
+# In these cases we can just `convert` as we know we are dealing with plain and simple types
+(::ProjectTo{T})(dx::AbstractFloat) where T<:AbstractFloat = convert(T, dx)
+(::ProjectTo{T})(dx::Integer) where T<:AbstractFloat = convert(T, dx)  #needed to avoid ambiguity
+
+# We asked for a number/real and they gave use one. We did ask for a particular concrete
+# type, but that is just for the preserving low precision floats, which is handled above.
+# Any Number/Real actually occupies the same subspace, so we can trust them.
+# In particular, this makes weirder Real subtypes that are not simply the values like
+# ForwardDiff.Dual and Symbolics.Sym work, because we stay out of their way.
+(::ProjectTo{<:Number})(dx::Number) = dx 
+# If you remove the above julia sometimes can't find the (::ProjectTo{T})(::T) for complex T
+# Seems like it might be a julia bug?
+
+(project::ProjectTo{<:Real})(dx::Complex) = project(real(dx))
```

Comment:
Can we trim this essay?
```suggestion
# Other numbers, including e.g. ForwardDiff.Dual and Symbolics.Sym, should pass through.
# (Unclear why the (::ProjectTo{T})(::T) method doesn't just work, for complex T)
(::ProjectTo{<:Number})(dx::Number) = dx 

(project::ProjectTo{<:Real})(dx::Complex) = project(real(dx))
```

Reply:
```suggestion
# Other numbers, including e.g. ForwardDiff.Dual and Symbolics.Sym, should pass through.
# We assume (lacking evidence to the contrary) that 
# The (::ProjectTo{T})(::T) method doesn't work because we are allowing a different
# Number type that might not be a subtype of the `project_type`.
(::ProjectTo{<:Number})(dx::Number) = dx 

(project::ProjectTo{<:Real})(dx::Complex) = project(real(dx))
```